src/bootstrap/src/utils/helpers.rs

```diff
@@ -150,6 +150,17 @@ pub fn symlink_dir(config: &Config, original: &Path, link: &Path) -> io::Result<
     }
 }
 
+/// Rename a file if from and to are in the same filesystem or
+/// copy and remove the file otherwise
+pub fn move_file<P: AsRef<Path>, Q: AsRef<Path>>(from: P, to: Q) -> io::Result<()> {
+    if fs::rename(&from, &to).is_err() {
```

Comment:
This is too general error handling, can this match on exact error kind? Or cfg(target).

Reply:
Do we have an ErrorKind for "Invalid cross-device link" ? If we do not it is really annoying to be more precise to spare calling copy that then would error out with the same error as rename.

Reply:
https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.CrossesDevices ?

Reply:
Nightly only.

Reply:
Can be done like here
https://github.com/rust-lang/rust/blob/dde8cfa597dff58e5f65f1184a83c2cc22318b6e/src/bootstrap/src/core/build_steps/clean.rs#L204-L211

libc::EXDEV is 18;

Reply:
Your impl will ignore errors except `EXDEV`, which is wrong; errors should be forwarded, similar to linked example.

Should function still be named `rename`? It will use rename in most cases, using copy&remove in fallback ones.

Reply:
it is mainly used to move files around, otherwise we'd not have the issue to begin with :)